### ISSUE TITLE
`impl` types lose information

### ISSUE THREAD

**Issue by LoganDark** [intellij-rust/intellij-rust#8853]
## Environment

* **IntelliJ Rust plugin version:** 0.4.171.4617-221-nightly
* **Rust toolchain version:** 1.62.0-nightly (88860d547 2022-05-09) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2022.1 (CL-221.5080.224)
* **Operating system:** Windows 10 Home 2004 build 19041.928 (64-bit)
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.emulate.terminal

## Problem description
IntelliJ-Rust does not see any issues with this code (based on #5410):

```rust
fn foo(a: impl AsRef<str>) {
    bar(a);
}

fn bar(_: impl AsRef<str> + Copy) {}
```

IntelliJ-Rust cannot infer the value of `s` for this code:

```rust
fn foo(x: impl AsRef<str>) {
    let s = x.as_ref();
}
```

I think this is because information is lost:

![image](https://user-images.githubusercontent.com/4723091/169199050-e90e34cc-205f-4155-9c4d-8c8480a41afb.png)

## Steps to reproduce
N/A combined into above

**Comment by Undin** [comment]
Can't reproduce either with `0.4.171.4630-221-nightly` or with `0.4.170.4627-221` plugin builds

**Comment by LoganDark** [comment]
> Can't reproduce either with `0.4.171.4630-221-nightly` or with `0.4.170.4627-221` plugin builds

What is it that you can't reproduce? Types showing up only as `impl`, or the type inference not working properly? Have you enabled all the experimental features listed in my issue? FWIW I am testing these examples in a scratch file.

**Comment by Undin** [comment]
> What is it that you can't reproduce? Types showing up only as impl, or the type inference not working properly?

Type inference works as expected
<img width="316" alt="image" src="https://user-images.githubusercontent.com/2539310/169480786-ee80b958-974b-4edc-843c-1621e1ced04d.png">

> Have you enabled all the experimental features listed in my issue?

`org.rust.cargo.emulate.terminal` doesn't affect type inference so it doesn't matter

> I am testing these examples in a scratch file. Does analysis not work at all there?

Oh, it's important part. Analysis partially works in scratch files, but since there is no info about stdlib in scratch files (you can check it by trying to navigate to `AsRef`. It shouldn't work currently), it's impossible to understand what `as_ref` does

**Comment by Undin** [comment]
Let's keep it open to check all known issues when we will improve support for scratch files

**Comment by Undin** [comment]
BTW @dima74 it looks quite simple now since we have an ability to dynamically create "fake" crates like we do for doc comments. And if I understand correctly, we just need to create such fake crate for each scratch file, attach dependencies on any known stdlib and all common features will work as expected

**Comment by LoganDark** [comment]
> BTW @dima74 it looks quite simple now since we have an ability to dynamically create "fake" crates like we do for doc comments. And if I understand correctly, we just need to create such fake crate for each scratch file, attach dependencies on any known stdlib and all common features will work as expected

Would be nice if I could configure which target is used for a scratch file then. My host OS is Windows but I develop crates in WSL.

**Comment by Undin** [comment]
> Would be nice if I could configure which target is used for a scratch file then. My host OS is Windows but I develop crates in WSL.

Yeah, in ideal case, the plugin should allow configuring what real workspace is considered as "root" one, i.e. where toolchain settings, stdlib path, cfg options and other things should be taken from. But in most cases, there is only one workspace in a project. In combination with the fact that plugin can't work properly with the case when different workspaces have different settings, the simplest solution (take all settings from any existing cargo workspace) should work in almost all cases

**Comment by LoganDark** [comment]
> take all settings from any existing cargo workspace

I have multiple crates in 1 project that are not part of the same workspace, this will still work for that, right?

**Comment by Undin** [comment]
> I have multiple crates in 1 project that are not part of the same workspace, this will still work for that, right?

It works if you have similar settings (toolchain version, build target, etc.) for all crates. If you try to have, for example, different toolchain versions for different crates (for instance, via toolchain override), it may work unpredictably in some cases
So, assuming you have similar settings, the described solution should work for scratch files even when project has several workspaces
